### ISSUE TITLE
FFI segfault on CFUNC :easy_cleanup

### ISSUE THREAD

**Issue by Zapotek** [typhoeus/typhoeus#492]
Got this on program exit after a Ctrl+C trap:
https://gist.github.com/Zapotek/0b74f29f07047de9b9cc

I think it may be caused by multiple blocking requests from different threads being aborted at the same time and the overlapping easy cleanups causing a segfault (these get their own Hydras right?).

```
curl 7.43.0 (x86_64-pc-linux-gnu) libcurl/7.43.0 GnuTLS/3.3.15 zlib/1.2.8 libidn/1.28 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP UnixSockets 
```

`:nosignal` and `:forbid_reuse` were enabled for the requests.

I was able reproduce it manually for a while but then it went away so I don't have a reliable test case.
(Using Typhoeus and Ethon from master branches.)

**Comment by i0rek** [comment]
> these get their own Hydras right?

You should have a hydra per thread. If you create multiple threads you need to take care of that on your own if you are using the hydra. Typhoeus tries to help a bit: https://github.com/typhoeus/typhoeus/blob/master/lib/typhoeus/hydra.rb#L67 but it is your responsibility in the end.

**Comment by Zapotek** [comment]
Yeah I'm using `Request#run` so Typhoeus must be using a Hydra per thread via the code you linked.
What's my responsibility though? Is there something I need to do to prevent those cleanup segfaults?

**Comment by i0rek** [comment]
When using `Request#run` it is not yours and not Typhoeus'. Then it is libcurls responsibility.

**Comment by Zapotek** [comment]
Well, at least it happens on exit, so that's good as far as segfaults go.
Did you spot anything interesting in the dump I linked? Hints or possibly Ruby workarounds?

**Comment by aliaksandrb** [comment]
Hey! This issue is still actual (as for typhoeus 1.3.1 / ethos 0.12.0).

I've see it quite often but not 100%, when typhoeus used in Rails environment, usually "at_exit" time (like console closed, job finished, rake task finished).
Seems to be GC related?

And it looks like if we provide GC callback to easy handle as block not method it fix the problem:
```ruby
# @handle ||= FFI::AutoPointer.new(Curl.easy_init, Curl.method(:easy_cleanup)) # current implementation
@handle ||= FFI::AutoPointer.new(Curl.easy_init, proc { |pointer| Curl.easy_cleanup(pointer) }) # works for me
```

Is it possible that `Curl.method(:easy_cleanup)` got binded to some context which is not available after GC started? Or it is somehow related to Rails-autoload feature (as I wasn't able to make POC outside of Rails).

I don't use hydra, only Request object with on_progress callback on it. And typhoeus used to make requests both, in "main" thread of application, and in background thread.

The other approach which seems to be working too is about synchronization of the call to `easy_cleanup`, here is an example: https://github.com/ffi/ffi/issues/447

Any suggestions how to address it better? While those segfaults are let the main task to be finished, it is kinda annoying and pollutes all logs / bug notification systems :)

**Comment by ericcj** [comment]
even with the latest versions:

typhoeus (1.4.0)
ethon (0.12.0)
ffi (1.13.1)

we also have this problem.  for us it is triggered by an exception within typhoeus:  https://github.com/typhoeus/typhoeus/issues/657#issuecomment-670926136
